src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+  }
+
+  override fun onCommissioningComplete(nodeId: Long, errorCode: Int) {
+    completionListener?.onCommissioningComplete(nodeId, errorCode)
+  }
+
+  override fun onCommissioningStatusUpdate(nodeId: Long, stage: String?, errorCode: Int) {
+    completionListener?.onCommissioningStatusUpdate(nodeId, stage, errorCode)
+  }
+
+  override fun onReadCommissioningInfo(
+    vendorId: Int,
+    productId: Int,
+    wifiEndpointId: Int,
+    threadEndpointId: Int
+  ) {
+    completionListener?.onReadCommissioningInfo(
+      vendorId,
+      productId,
+      wifiEndpointId,
+      threadEndpointId
+    )
+  }
+
+  override fun onOpCSRGenerationComplete(csr: ByteArray) {
+    completionListener?.onOpCSRGenerationComplete(csr)
+  }
+
+  override fun onError(error: Throwable) {
+    completionListener?.onError(error)
+  }
+
+  override fun subscribe(request: SubscribeRequest): Flow<SubscriptionState> {
+    // To prevent potential issues related to concurrent modification, assign
+    // the value of the mutable property 'nodeId' to a temporary variable.
+    val nodeId = this.nodeId
+    check(nodeId != null) { "nodeId has not been initialized yet" }
+
+    val attributePaths = generateAttributePaths(request)
+    val eventPaths = generateEventPaths(request)
+    val successes = mutableListOf<ReadData>()
+    val failures = mutableListOf<ReadFailure>()
+
+    return callbackFlow {
+        val devicePtr: Long = getConnectedDevicePointer(nodeId)
+        val subscriptionEstablishedHandler = SubscriptionEstablishedCallback {
+          logger.log(Level.INFO, "Subscription to device established")
+
+          trySendBlocking(SubscriptionState.SubscriptionEstablished).onFailure { ex ->
+            logger.log(
+              Level.SEVERE,
+              "Error sending SubscriptionCompletedNotification to subscriber: %s",
+              ex
+            )
+          }
+        }
+
+        val resubscriptionAttemptHandler =
+          ResubscriptionAttemptCallback { terminationCause, nextResubscribeIntervalMsec ->
+            logger.log(
+              Level.WARNING,
+              "ResubscriptionAttempt terminationCause:${terminationCause}, " +
+                "nextResubscribeIntervalMsec:${nextResubscribeIntervalMsec}"
+            )
+
+            trySendBlocking(
+                SubscriptionState.SubscriptionErrorNotification(terminationCause.toUInt())
+              )
+              .onFailure { ex ->
+                logger.log(
+                  Level.SEVERE,
+                  "Error sending ResubscriptionNotification to subscriber: %s",
+                  ex
+                )
+              }
+          }
+
+        val reportHandler =
+          object : ReportCallback {
+            override fun onReport(nodeState: chip.devicecontroller.model.NodeState) {
+              logger.log(Level.INFO, "Received subscribe update report")
+
+              val tmpNodeState: NodeState = nodeState.wrap()
```

Comment:
why do we require separate classes to wrap these? Since we place data directly into success/failures without considering state, having a copy of state elements seems to make things somewhat more complex.

Reply:
This would not be required if we had direct jni calls, but we do need it right now, since the user of this API does not see the underneath class type from Java layer in SDK.